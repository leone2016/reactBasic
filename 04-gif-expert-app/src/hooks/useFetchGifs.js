import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    /**
     * useEffect: se utiliza para evitar que se realice una peticion fetch
     * cada que el estado del componente cambia, ademas
     * Los efectos NO puedes ser async 
     * SOLUCION, add in the [ category ]
     * 
     * React Hook useEffect has a missing dependency: 'getGifs'. 
     * Either include it or remove the dependency array  react-hooks/exhaustive-deps
     */
    useEffect(() => {
        getGifs(category).then(imgs => {
            setState({
                data: imgs,
                loading: false
            })
        });
    }, [category]);



    return state;
}
