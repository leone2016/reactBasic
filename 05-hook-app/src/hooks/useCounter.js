import { useState } from "react"
/**
 * TEst
 * @param {*} initialState 
 * @returns {state,
        increment,
        decrement,
        resetear}
 */
export const useCounter = (initialState = 10 ) => {
    const [state, setstate] = useState(initialState);

    const increment = (factor = 1)=>{
        setstate( state + factor);
    }
    const decrement = (factor = 1)=>{
        setstate( state - factor);
    }
    const resetear = ()=>{
        setstate( initialState );
    }

    //useCounter
    return {
        state,
        increment,
        decrement,
        resetear
    }
}
