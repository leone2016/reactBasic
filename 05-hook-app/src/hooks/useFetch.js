import  { useState, useEffect, useRef } from 'react'

export const useFetch = (url) => {

    const isMounted = useRef(true);
    const initialState = { data: null, loading:true, error: null};
    const [state, setstate] = useState(initialState);

    useEffect(() => {
        return () => {
            isMounted.current = false; // este cambio no realiza la renderizacion del componente, solo la referencia del mismo 
        }
    }, [])


    useEffect(() => {
        setstate(initialState);
        fetch(url)
        .then( resp => resp.json() )
        .then( data => {
            setTimeout(() =>{
                if(isMounted.current){
                    setstate({ 
                        loading: false,
                        error: null,
                        data
                    });
                }else{
                    console.warn('setState no se llamo');
                }
                

            },4000);
           
        })
    },[url]);

    return  state;
}
