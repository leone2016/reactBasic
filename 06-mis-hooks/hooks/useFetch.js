import  { useState, useEffect, useRef } from 'react'

export const useFetch = (url) => {

    const isMounted = useRef(true);
    const initialState = { data: null, loading:true, error: null};
    const [state, setstate] = useState(initialState);

    useEffect(() => {
        return () => {
            isMounted.current = false; // este cambio no realiza la renderizacion del componente, solo la referencia del mismo 
        }
    }, [])


    useEffect(() => {
        setstate(initialState);
        fetch(url)
        .then( resp => resp.json() )
        .then( data => {
           /*  setTimeout(() =>{ */
                if(isMounted.current){
                    setstate({ 
                        loading: false,
                        error: null,
                        data
                    });
                }
                

           /*  },4000); */
           
        }).catch(()=>{
            setstate({
                data: null, 
                loading: false, 
                error: 'No se pudo cargar la info'
            });
        })
    },[url]);

    return  state;
}
