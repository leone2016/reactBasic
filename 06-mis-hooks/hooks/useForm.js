import  { useState } from 'react'
/**
 * Permite manejar cualquier formulario
 * 
 * @param {*} initialState  un ojeto de cualquier tipo
 */
export const useForm = (initialState = {}) => {

    const [values, setvalues] = useState(initialState);

    const handleInputChange = ({ target }) => {
        setvalues({
            ...values,
            [target.name]: target.value
        })
    }

    const reset = () => {
        setvalues(initialState);
    }

    return [values, handleInputChange, reset ];

}
