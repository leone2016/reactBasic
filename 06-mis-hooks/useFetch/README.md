# useFetch

ejemplo de use:

````Js
    /**
     * 
     */
    const url = 'endPoint_de_una_app';
    const { data, loading, error} = useFetch(url);
````