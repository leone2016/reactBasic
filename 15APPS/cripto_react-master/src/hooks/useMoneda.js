import React, {Fragment, useState } from 'react';
import styled from '@emotion/styled';

const Label = styled.label`
    font-family: 'Bebas Neue', cursive;
    color: #FFF;
    text-transform: uppercase;
    font-weight: bold;
    font-size: 2.4rem;
    margin-top: 2rem;
    display: block; /*toma todo el espacio que tiene disponible y evita que se imprima en la misma linea */
`;

const Select = styled.select`
    width: 100%;
    display:block;
    padding: 1rem;
    -webkit-appearance: none;
    border-radius: 10px;
    border: none;
    font-size: 1.2rem;
`

const useMoneda = (label, stateInicial, opciones) => {

    // State de nuestro custom hook
    const [state, actualizarState] = useState(stateInicial);

    /**
     * Esto sera lo que se muestre en pantalla
     * @returns {JSX.Element}
     * @constructor
     */
    const Seleccionar = () => (
        <Fragment>
            <Label>{label}</Label>
            <Select
                onChange={ e => actualizarState(e.target.value)}
                value={state}
            >
                <option value="">- Seleccione -</option>
                {opciones.map(opcion => (
                    <option key={opcion.codigo} value={opcion.codigo}>{opcion.nombre}</option>
                ))}
            </Select>
        </Fragment>
    );

    // Retornar state, interfaz y fn que modifica el state
    return [state, Seleccionar, actualizarState];
}

export default useMoneda;
