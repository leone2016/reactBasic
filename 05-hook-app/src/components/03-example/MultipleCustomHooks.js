import React from 'react'
import { useFetch } from '../../hooks/useFetch'
import {useCounter} from '../../hooks/useCounter'
/**
 * sirve para que varios hooks interactuen entre si 
 */
export const MultipleCustomHooks = () => {

    const {state:counter, increment } = useCounter(1);
    const stateFetch = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
    console.log(stateFetch);
    const { loading, data } = stateFetch;
    const { author, quote } = !!data && data[0]; // !!data: Doble degacion, 1ra: !null 
    return (
        <>
            <h1> Breaking Bad Quotes  </h1>
            <hr />
            {
                loading ? (<div className="loading" >L O A D I N G</div>)
                    :
                    (<blockquote className="blockquote text-right">
                        <p className="mb-0">{quote}</p>
                        <footer className="blockquote-footer">{author}</footer>
                    </blockquote>)

            }

<button onClick={ () => increment(1)}>Siguiente</button>


        </>
    )
}
