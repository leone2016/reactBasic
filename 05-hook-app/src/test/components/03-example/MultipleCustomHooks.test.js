import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-example/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';
// el jest.mock lo que hace que cuando se vaya a utilizar ese archivo, en vez de utilizar el useFetch directo,
// utiliza una implementaci√≥n
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Pruebas en <MultipleCustomHoks/>', () => {

    useCounter.mockReturnValue({ 
        state: 10, 
        increment: ( ) => {},
     });

    test('debe de mostrar correctamente', () => {
        // esto es lo que retorna por defecto el useFetch
        useFetch.mockReturnValue({
            data: null, loading: true, error: null
        })
        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de mostrar la informacion', () => {
        useFetch.mockReturnValue({
            data: [{
                author: 'Leonardo',
                quote: 'Hola mundo 2020'
            }],
            loading: false,
            error: null
        })
        const wrapper = shallow(<MultipleCustomHooks />);
        /* console.log(wrapper.html()); */
        expect(wrapper.find('.loading').exists()).toBe( false ); // cuando tenga data, no debe existir la clase .loading
        expect(wrapper.find('.mb-0').text().trim()).toBe( 'Hola mundo 2020' ); // la clase .mb-0 <footer/>
        expect(wrapper.find('footer').text().trim()).toBe( 'Leonardo' ); // la etiqueta <footer/>
    });

})
