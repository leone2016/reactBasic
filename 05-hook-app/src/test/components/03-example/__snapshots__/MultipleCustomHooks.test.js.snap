// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHoks/> debe de mostrar correctamente 1`] = `
<Fragment>
  <h1>
     Breaking Bad Quotes  
  </h1>
  <hr />
  <div
    className="loading"
  >
    L O A D I N G
  </div>
  <button
    onClick={[Function]}
  >
    Siguiente
  </button>
</Fragment>
`;
